GLMakie/src/glwindow.jl

```diff
@@ -10,182 +10,104 @@ end
 Base.convert(::Type{SelectionID{T}}, s::SelectionID) where T = SelectionID{T}(T(s.id), T(s.index))
 Base.zero(::Type{GLMakie.SelectionID{T}}) where T = SelectionID{T}(T(0), T(0))
 
-mutable struct GLFramebuffer
-    resolution::Observable{NTuple{2, Int}}
-    id::GLuint
 
-    buffer_ids::Dict{Symbol, GLuint}
-    buffers::Dict{Symbol, Texture}
-    render_buffer_ids::Vector{GLuint}
-end
-
-# it's guaranteed, that they all have the same size
-Base.size(fb::GLFramebuffer) = size(fb.buffers[:color])
-Base.haskey(fb::GLFramebuffer, key::Symbol) = haskey(fb.buffers, key)
-Base.getindex(fb::GLFramebuffer, key::Symbol) = fb.buffer_ids[key] => fb.buffers[key]
-
-function getfallback(fb::GLFramebuffer, key::Symbol, fallback_key::Symbol)
-    haskey(fb, key) ? fb[key] : fb[fallback_key]
-end
-
-
-function attach_framebuffer(t::Texture{T, 2}, attachment) where T
-    glFramebufferTexture2D(GL_FRAMEBUFFER, attachment, GL_TEXTURE_2D, t.id, 0)
-end
-
-# attach texture as color attachment with automatic id picking
-function attach_colorbuffer!(fb::GLFramebuffer, key::Symbol, t::Texture{T, 2}) where T
-    if haskey(fb.buffer_ids, key) || haskey(fb.buffers, key)
-        error("Key $key already exists.")
-    end
-
-    max_color_id = GL_COLOR_ATTACHMENT0
-    for id in values(fb.buffer_ids)
-        if GL_COLOR_ATTACHMENT0 <= id <= GL_COLOR_ATTACHMENT15 && id > max_color_id
-            max_color_id = id
-        end
-    end
-    next_color_id = max_color_id + 0x1
-    if next_color_id > GL_COLOR_ATTACHMENT15
-        error("Ran out of color buffers.")
-    end
 
-    glFramebufferTexture2D(GL_FRAMEBUFFER, next_color_id, GL_TEXTURE_2D, t.id, 0)
-    push!(fb.buffer_ids, key => next_color_id)
-    push!(fb.buffers, key => t)
-    return next_color_id
-end
+mutable struct FramebufferFactory
+    fb::GLFramebuffer # core framebuffer (more or less for #4150)
```

Comment:
We need a better name for this... What's this doing exactly?

Reply:
I read that GLFramebuffers are very cheap to build/use and it's generally advised to have many dedicated ones over a few everything-framebuffers.  So the design here is that the Makie side figures out which buffers need to be in a given render stage and the GLMakie side then builds them accordingly. This object is responsible for building them (which includes reuse of attachments) and keeping track of all the attachments and framebuffers for resizing and cleanup. 

Maybe calling this a factory is wrong and should be something like `FramebufferManager`? Though the main job/complexity is building GLFramebuffer objects.

Reply:
Yeah, sounds more like a manager....